// victory-core/lib/victory-primitives/voronoi.js
Object.defineProperty(exports, "__esModule", {
  value: true
});

var _uniqueId2 = require("lodash/uniqueId");

var _uniqueId3 = _interopRequireDefault(_uniqueId2);

var _isObject2 = require("lodash/isObject");

var _isObject3 = _interopRequireDefault(_isObject2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _helpers = require("../victory-util/helpers");

var _helpers2 = _interopRequireDefault(_helpers);

var _commonProps = require("./common-props");

var _commonProps2 = _interopRequireDefault(_commonProps);

var _clipPath = require("../victory-primitives/clip-path");

var _clipPath2 = _interopRequireDefault(_clipPath);

var _path = require("../victory-primitives/path");

var _path2 = _interopRequireDefault(_path);

var _circle = require("../victory-primitives/circle");

var _circle2 = _interopRequireDefault(_circle);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /*eslint no-magic-numbers: ["error", { "ignore": [2] }]*/


var Voronoi = function (_React$Component) {
  _inherits(Voronoi, _React$Component);

  function Voronoi(props) {
    _classCallCheck(this, Voronoi);

    var _this = _possibleConstructorReturn(this, (Voronoi.__proto__ || Object.getPrototypeOf(Voronoi)).call(this, props));

    _this.clipId = !(0, _isObject3.default)(props) || typeof props.clipId === "undefined" ? (0, _uniqueId3.default)("voronoi-clip-") : props.clipId;
    return _this;
  }

  _createClass(Voronoi, [{
    key: "getVoronoiPath",
    value: function getVoronoiPath(props) {
      var polygon = props.polygon;

      return Array.isArray(polygon) && polygon.length ? "M " + props.polygon.join("L") + " Z" : "";
    }
  }, {
    key: "render",
    value: function render() {
      var _props = this.props,
          datum = _props.datum,
          active = _props.active,
          role = _props.role,
          shapeRendering = _props.shapeRendering,
          className = _props.className,
          events = _props.events,
          x = _props.x,
          y = _props.y,
          pathComponent = _props.pathComponent,
          clipPathComponent = _props.clipPathComponent,
          groupComponent = _props.groupComponent,
          circleComponent = _props.circleComponent;

      var voronoiPath = this.getVoronoiPath(this.props);
      var style = _helpers2.default.evaluateStyle(this.props.style, datum, active);
      var size = _helpers2.default.evaluateProp(this.props.size, datum, active);

      if (size) {
        var circle = _react2.default.cloneElement(circleComponent, {
          key: "circle", style: style, className: className, role: role, shapeRendering: shapeRendering, events: events,
          clipPath: "url(#" + this.clipId + ")", cx: x, cy: y, r: size
        });
        var voronoiClipPath = _react2.default.cloneElement(clipPathComponent, { key: "voronoi-clip", clipId: this.clipId }, _react2.default.cloneElement(pathComponent, { d: voronoiPath, className: className }));
        return _react2.default.cloneElement(groupComponent, {}, [voronoiClipPath, circle]);
      }
      return _react2.default.cloneElement(pathComponent, {
        style: style, className: className, d: voronoiPath, role: role, shapeRendering: shapeRendering, events: events
      });
    }
  }]);

  return Voronoi;
}(_react2.default.Component);

Voronoi.propTypes = _extends({}, _commonProps2.default, {
  circleComponent: _propTypes2.default.element,
  clipId: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]),
  clipPathComponent: _propTypes2.default.element,
  datum: _propTypes2.default.object,
  groupComponent: _propTypes2.default.element,
  pathComponent: _propTypes2.default.element,
  polygon: _propTypes2.default.array,
  size: _propTypes2.default.number,
  x: _propTypes2.default.number,
  y: _propTypes2.default.number
});
Voronoi.defaultProps = {
  pathComponent: _react2.default.createElement(_path2.default, null),
  circleComponent: _react2.default.createElement(_circle2.default, null),
  clipPathComponent: _react2.default.createElement(_clipPath2.default, null),
  groupComponent: _react2.default.createElement("g", null)
};
exports.default = Voronoi;
