// victory-chart/lib/components/victory-boxplot/victory-boxplot.js
Object.defineProperty(exports, "__esModule", {
  value: true
});

var _flatten2 = require("lodash/flatten");

var _flatten3 = _interopRequireDefault(_flatten2);

var _partialRight2 = require("lodash/partialRight");

var _partialRight3 = _interopRequireDefault(_partialRight2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _commonProps = require("../../helpers/common-props");

var _victoryCore = require("victory-core");

var _helperMethods = require("./helper-methods");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var fallbackProps = {
  width: 450,
  height: 300,
  padding: {
    top: 20,
    right: 20,
    bottom: 20,
    left: 20
  }
};

var options = {
  components: [{ name: "min" }, { name: "minLabels" }, { name: "max" }, { name: "maxLabels" }, { name: "median" }, { name: "medianLabels" }, { name: "q1" }, { name: "q1Labels" }, { name: "q3" }, { name: "q3Labels" }, { name: "parent", index: "parent" }]
};

var defaultData = [{ x: 1, min: 5, q1: 7, median: 12, q3: 18, max: 20 }, { x: 2, min: 2, q1: 5, median: 8, q3: 12, max: 15 }];

var animationWhitelist = ["data", "domain", "height", "padding", "style", "width"];

var VictoryBoxPlot = function (_React$Component) {
  _inherits(VictoryBoxPlot, _React$Component);

  function VictoryBoxPlot() {
    _classCallCheck(this, VictoryBoxPlot);

    return _possibleConstructorReturn(this, (VictoryBoxPlot.__proto__ || Object.getPrototypeOf(VictoryBoxPlot)).apply(this, arguments));
  }

  _createClass(VictoryBoxPlot, [{
    key: "renderBoxPlot",
    value: function renderBoxPlot(props) {
      var _this2 = this;

      var types = ["q1", "q3", "max", "min", "median"];
      var dataComponents = (0, _flatten3.default)(types.map(function (type) {
        return _this2.dataKeys.map(function (key, index) {
          var baseComponent = props[type + "Component"];
          var componentProps = _this2.getComponentProps(baseComponent, type, index);
          return _react2.default.cloneElement(baseComponent, componentProps);
        });
      }));

      var labelComponents = (0, _flatten3.default)(types.map(function (type) {
        var components = _this2.dataKeys.map(function (key, index) {
          var name = type + "Labels";
          var baseComponent = props[type + "LabelComponent"];
          var labelProps = _this2.getComponentProps(baseComponent, name, index);
          if (typeof labelProps.text !== "undefined" && labelProps.text !== null) {
            return _react2.default.cloneElement(baseComponent, labelProps);
          }
          return undefined;
        });
        return components.filter(Boolean);
      }));
      var children = [].concat(_toConsumableArray(dataComponents), _toConsumableArray(labelComponents));
      return this.renderContainer(props.groupComponent, children);
    }

    // Overridden in native versions

  }, {
    key: "shouldAnimate",
    value: function shouldAnimate() {
      return !!this.props.animate;
    }
  }, {
    key: "render",
    value: function render() {
      var role = this.constructor.role;

      var props = _victoryCore.Helpers.modifyProps(this.props, fallbackProps, role);
      if (this.shouldAnimate()) {
        return this.animateComponent(props, animationWhitelist);
      }
      var children = this.renderBoxPlot(props);
      return props.standalone ? this.renderContainer(props.containerComponent, children) : children;
    }
  }]);

  return VictoryBoxPlot;
}(_react2.default.Component);

VictoryBoxPlot.displayName = "VictoryBoxPlot";
VictoryBoxPlot.role = "boxplot";
VictoryBoxPlot.defaultTransitions = _victoryCore.DefaultTransitions.discreteTransitions();
VictoryBoxPlot.propTypes = _extends({}, _commonProps.BaseProps, _commonProps.DataProps, {
  boxWidth: _propTypes2.default.number,
  events: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    target: _propTypes2.default.oneOf(["max", "maxLabels", "median", "medianLabels", "min", "minLabels", "q1", "q1Labels", "q3", "q3Labels", "parent"]),
    eventKey: _propTypes2.default.oneOfType([_propTypes2.default.array, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string]),
    eventHandlers: _propTypes2.default.object
  })),
  horizontal: _propTypes2.default.bool,
  labelOrientation: _propTypes2.default.oneOf(["top", "bottom", "left", "right"]),
  labels: _propTypes2.default.bool,
  max: _propTypes2.default.oneOfType([_propTypes2.default.func, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  maxComponent: _propTypes2.default.element,
  maxLabelComponent: _propTypes2.default.element,
  maxLabels: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.array, _propTypes2.default.bool]),
  median: _propTypes2.default.oneOfType([_propTypes2.default.func, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  medianComponent: _propTypes2.default.element,
  medianLabelComponent: _propTypes2.default.element,
  medianLabels: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.array, _propTypes2.default.bool]),
  min: _propTypes2.default.oneOfType([_propTypes2.default.func, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  minComponent: _propTypes2.default.element,
  minLabelComponent: _propTypes2.default.element,
  minLabels: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.array, _propTypes2.default.bool]),
  q1: _propTypes2.default.oneOfType([_propTypes2.default.func, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  q1Component: _propTypes2.default.element,
  q1LabelComponent: _propTypes2.default.element,
  q1Labels: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.array, _propTypes2.default.bool]),
  q3: _propTypes2.default.oneOfType([_propTypes2.default.func, _victoryCore.PropTypes.allOfType([_victoryCore.PropTypes.integer, _victoryCore.PropTypes.nonNegative]), _propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  q3Component: _propTypes2.default.element,
  q3LabelComponent: _propTypes2.default.element,
  q3Labels: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.array, _propTypes2.default.bool]),
  style: _propTypes2.default.shape({
    boxes: _propTypes2.default.object,
    labels: _propTypes2.default.object,
    parent: _propTypes2.default.object,
    max: _propTypes2.default.object,
    maxLabels: _propTypes2.default.object,
    median: _propTypes2.default.object,
    medianLabels: _propTypes2.default.object,
    min: _propTypes2.default.object,
    minLabels: _propTypes2.default.object,
    q1: _propTypes2.default.object,
    q1Labels: _propTypes2.default.object,
    q3: _propTypes2.default.object,
    q3Labels: _propTypes2.default.object,
    whiskers: _propTypes2.default.object
  }),
  whiskerWidth: _propTypes2.default.number
});
VictoryBoxPlot.defaultProps = {
  containerComponent: _react2.default.createElement(_victoryCore.VictoryContainer, null),
  data: defaultData,
  dataComponent: _react2.default.createElement(_victoryCore.Box, null),
  groupComponent: _react2.default.createElement("g", { role: "presentation" }),
  maxComponent: _react2.default.createElement(_victoryCore.Whisker, null),
  maxLabelComponent: _react2.default.createElement(_victoryCore.VictoryLabel, null),
  medianComponent: _react2.default.createElement(_victoryCore.Line, null),
  medianLabelComponent: _react2.default.createElement(_victoryCore.VictoryLabel, null),
  minComponent: _react2.default.createElement(_victoryCore.Whisker, null),
  minLabelComponent: _react2.default.createElement(_victoryCore.VictoryLabel, null),
  q1Component: _react2.default.createElement(_victoryCore.Box, null),
  q1LabelComponent: _react2.default.createElement(_victoryCore.VictoryLabel, null),
  q3Component: _react2.default.createElement(_victoryCore.Box, null),
  q3LabelComponent: _react2.default.createElement(_victoryCore.VictoryLabel, null),
  samples: 50,
  scale: "linear",
  sortKey: "x",
  sortOrder: "ascending",
  standalone: true,
  theme: _victoryCore.VictoryTheme.grayscale
};
VictoryBoxPlot.getDomain = _helperMethods.getDomain;
VictoryBoxPlot.getData = _helperMethods.getData;
VictoryBoxPlot.getBaseProps = (0, _partialRight3.default)(_helperMethods.getBaseProps, fallbackProps);
VictoryBoxPlot.expectedComponents = ["maxComponent", "maxLabelComponent", "medianComponent", "medianLabelComponent", "minComponent", "minLabelComponent", "q1Component", "q1LabelComponent", "q3Component", "q3LabelComponent", "groupComponent", "containerComponent"];
exports.default = (0, _victoryCore.addEvents)(VictoryBoxPlot, options);
